src/Stack/SDist.hs

```diff
@@ -93,71 +95,98 @@ instance Show CheckException where
-    -- for upload (both GZip.compress and Tar.write are lazy).
-    -- However, it seems less error prone and more predictable to read
-    -- everything in at once, so that's what we're doing for now:
-    let tarPath isDir fp = either throwString return
-            (Tar.toTarPath isDir (forceUtf8Enc (pkgId FP.</> fp)))
-        -- convert a String of proper characters to a String of bytes
-        -- in UTF8 encoding masquerading as characters. This is
-        -- necessary for tricking the tar package into proper
-        -- character encoding.
-        forceUtf8Enc = S8.unpack . T.encodeUtf8 . T.pack
-        packWith f isDir fp = liftIO $ f (pkgFp FP.</> fp) =<< tarPath isDir fp
-        packDir = packWith Tar.packDirectoryEntry True
-        packFile fp
+    let getCabalFile fp fi
             -- This is a cabal file, we're going to tweak it, but only
             -- tweak it as a revision.
-            | tweakCabal && isCabalFp fp && asRevision = do
+            | tweakCabal && asRevision = do
                 lbsIdent <- getCabalLbs pvpBounds (Just 1) cabalfp sourceMap
                 liftIO (writeIORef cabalFileRevisionRef (Just lbsIdent))
-                packWith packFileEntry False fp
+                return (fi, C.sourceFile fp)
             -- Same, except we'll include the cabal file in the
             -- original tarball upload.
-            | tweakCabal && isCabalFp fp = do
+            | tweakCabal = do
                 (_ident, lbs) <- getCabalLbs pvpBounds Nothing cabalfp sourceMap
-                currTime <- liftIO getPOSIXTime -- Seconds from UNIX epoch
-                tp <- liftIO $ tarPath False fp
-                return $ (Tar.fileEntry tp lbs) { Tar.entryTime = floor currTime }
-            | otherwise = packWith packFileEntry False fp
+                curTime <- curModTime
+                let fi' =
+                        fi
+                            { CTar.fileModTime = curTime
+                            , CTar.fileSize = fromIntegral (L.length lbs)
+                            }
+                pure (fi', C.sourceLazy lbs)
+            | otherwise = pure (fi, C.sourceFile fp)
         isCabalFp fp = toFilePath pkgDir FP.</> fp == toFilePath cabalfp
-        tarName = pkgId FP.<.> "tar.gz"
         pkgId = packageIdentifierString (packageIdentifier (lpPackage lp))
-    dirEntries <- mapM packDir (dirsFromFiles files)
-    fileEntries <- mapM packFile files
+        yieldPackageFiles fs conf =
+            C.fuseUpstream
+                (C.yieldMany fs .| CTar.tarFiles conf)
+                (C.mapC (mapLeft resetFileInfo))
+        yieldCabalFile fp = do
+            fi <- resetFileInfo <$> CTar.getFileInfo fp
+            (fiAdjusted, cabalFileSource) <- lift $ lift $ getCabalFile fp fi
+            C.yield $
+                Left $
+                fiAdjusted
+                    { CTar.filePath =
+                          CTar.encodeFilePath (pkgId <> "/" <> FP.normalise fp)
```

Comment:
I think `CTar.getFileInfo` already normalizes the filepath so this could be redundant. Verify it and remove if that's the case